ipfs: initialConfig option + fixes

elitak opened the pull request:
added initialConfig option
fix: include wrapped binary in system PATH
fix: init service now attempts to umount mountpoints
fix: properly evaluate overridden extraConfig options.

###### Motivation for this change
fixes #42672

Addresses #82580 and ipfs/go-ipfs#4214 (https://github.com/NixOS/nixpkgs/pull/29133#issuecomment-613067876)

initialConfig option is useful for fresh deployments via nixops etc.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthewbauer commented:
@grahamcofborg test ipfs

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43089&attempt_id=89383086-3feb-42a4-9684-28dd5d1a5b66)

The following builds were skipped because they don't evaluate on aarch64-linux: tests.ipfs

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `tests.ipfs' because:
error: attribute 'ipfs' in selection path 'tests.ipfs' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43089&attempt_id=37a11171-d152-4348-bcaf-ca23738f5efa)

The following builds were skipped because they don't evaluate on x86_64-linux: tests.ipfs

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `tests.ipfs' because:
error: attribute 'ipfs' in selection path 'tests.ipfs' not found

```
</p></details>

matthewbauer commented:
/cc @seppeljordan

seppeljordan commented on nixos/modules/services/network-filesystems/ipfs.nix:
```diff
@@ -45,12 +50,14 @@ let
             ''
               read value <<EOF
-              ${builtins.toJSON value}
+              ${builtins.toJSON value'}
```
What is this apostrophe?

Infinisil commented on nixos/modules/services/network-filesystems/ipfs.nix:
```diff
@@ -45,12 +50,14 @@ let
-            (path: value:
+          (mapAttrsRecursiveCond (value: ! value ? _type)
+            (path: value: let
+              value' = if value ? _type then value.content else value;
```
Why is this suddenly needed?

elitak replied:
Attrsets in configuration.nix evaluate to these kind of attrset wrappers (for purposes of handling `mkOverride` IIUC), rather than the attrsets literally provided. I had not sufficiently tested using attrsets as values, so I only recently found this change was necessary for proper marshalling.

Infinisil replied:
I don't get it, can you add a comment explaining this with an example of what wouldn't work without it?

elitak replied:
`services.ipfs.extraConfig.Swarm.AddrFilters = [];` wouldn't work right now in master, because the attrset `Swarm` is not flattened into a simple attrset at the time the recursive map is evaluated. I can't explain why, because I don't understand it myself. I only know that the attrset I'm expecting is in `.contents` whenever `.type` is present.

Infinisil replied:
Can't reproduce, with or without this change, the mentioned setting works without problems.

mmahut replied:
Update on this?

Infinisil commented on nixos/modules/services/network-filesystems/ipfs.nix:
```diff
@@ -233,15 +254,24 @@ in {
-        install -m 0755 -o ${cfg.user} -g ${cfg.group} -d ${cfg.ipfsMountDir}
-        install -m 0755 -o ${cfg.user} -g ${cfg.group} -d ${cfg.ipnsMountDir}
+        # XXX Sometimes, the daemon stops or crashes without unmounting.
+        umount -l "${cfg.ipfsMountDir}" || true
```
Why a lazy unmount?

elitak replied:
If some process has an open handle in there, I don't want that to prevent a remount. This is a FUSE virtual FS, anyway, so I can't imagine how any harm could result from this.

mmahut commented:
Any updates on this pull request, please?

elitak committed:
ipfs: initialConfig option + fixes

fix: include wrapped binary in system PATH
fix: init service now attempts to umount mountpoints
fix: properly evaluate overridden extraConfig options.

elitak commented:
I've combined the original initialConfig featrue with a few other fixes. This is ready to be reviewed again.

Ericson2314 commented on nixos/modules/services/network-filesystems/ipfs.nix:
```diff
@@ -22,35 +27,45 @@ let
+      umount -l "${cfg.ipnsMountDir}" || true
+      install -m 0755 -o "${cfg.user}" -g "${cfg.group}" -d "${cfg.ipfsMountDir}"
+      install -m 0755 -o "${cfg.user}" -g "${cfg.group}" -d "${cfg.ipnsMountDir}"
+>>>>>>> 03a4ab9399b... ipfs: initialConfig option + fixes
```
merge artifact

mmahut commented:
@elitak can we please have this rebase with the fix?

Ericson2314 commented:
We want to land https://github.com/NixOS/nixpkgs/pull/90157/files, either with a fix, or a fix right after.

elitak commented:
Huh? This doesn't fix anything crucial. I can rebase it later.

Not sure what the hangup was over this PR; I opened it almost 2 years ago...

Ericson2314 commented:
Oh, I meant that I wasn't gonna rebase this myself. But fine if you do, we can merge it.

matthewbauer commented on nixos/modules/services/network-filesystems/ipfs.nix:
```diff
@@ -192,6 +207,20 @@ in {
         default = true;
       };
 
+      initialConfig = mkOption {
```
Why is initialConfig needed when extraConfig is also available?

matthewbauer commented on nixos/modules/services/network-filesystems/ipfs.nix:
```diff
@@ -22,33 +27,43 @@ let
     # NB: migration must be performed prior to pre-start, else we get the failure message!
-    preStart = optionalString cfg.autoMount ''
+    preStart = ''
+      install -m 0755 -o "${cfg.user}" -g "${cfg.group}" -d "${cfg.dataDir}"
```
is this still needed in ipfs 0.5.1? ideally we would get some more diagnostics and report this as a bug

Ericson2314 commented:
There is no wrapped binary, and I think the other 3 things would be better to configure separately. We have two bugs that cannot be reproduced, and `initialConfig` is a feature not a fix.

Ericson2314 commented:
I think we've taken everything we should from this.
